### PR TITLE
Move to RSS and HeapAlloc

### PR DESCRIPTION
**Pull request by gouthamve**
This is a little invasive, I'll integrate into the cortex-querier and update here.

### CODE REVIEW

**Comment by tomwilkie** [comment]
Tried this out on Prometheus today, LGTM!

File: cmd/demo-heap/main.go [hunk 1]
```diff
@@ -43,11 +43,17 @@ func LightAllocator() {
 func main() {
 	fmt.Println("allocating 1100 kB every second. should hit the 10MB threshold within 10 seconds.  look for a profile..")
 	errors := make(chan error)
-	trigger, _ := heap.New(".", 10000000, 60, time.Duration(1)*time.Second, errors)
+	trigger, _ := heap.New(heap.Config{
+		Path:           ".",
+		AllocThreshold: 10000000,
+		MinTimeDiff:    time.Duration(60) * time.Second,
+		CheckEvery:     time.Duration(1) * time.Second,
```
Comment (Dieterbe) [inline cmd/demo-heap/main.go]: don't need the `time.Duration()`. the type is inferred from the time.Second constant

File: cmd/demo-heap/main.go [hunk 2]
```diff
@@ -43,11 +43,17 @@ func LightAllocator() {
...
+	}, errors)
+
 	go trigger.Run()
 	go HungryAllocator()
 	go LightAllocator()
 	for e := range errors {
-		log.Fatal("profiletrigger heap saw error:", e)
+		log.Println("profiletrigger heap saw error:", e)
```
Comment (Dieterbe) [inline cmd/demo-heap/main.go]: why this change?

File: heap/heap.go [hunk 3]
```diff
@@ -6,6 +6,8 @@ import (
 	"runtime"
 	"runtime/pprof"
 	"time"
+
+	"github.com/prometheus/procfs"
```
Comment (Dieterbe) [inline heap/heap.go]: i don't want to pull in this dependency.
we've pulled that package in in some other projects and it's always a huge battle between dep and the fixtures directory which contains a bunch of symlinks which breaks a bunch of other tools.

can we just copy the small piece of code we need?

File: heap/heap.go [hunk 4]
```diff
@@ -14,43 +16,57 @@ import (
...
+func New(cfg Config, errors chan error) (*Heap, error) {
 	heap := Heap{
-		path,
-		threshold,
-		minTimeDiff,
-		checkEvery,
-		int64(0),
-		errors,
+		cfg:           cfg,
+		lastTriggered: time.Now().Add(-cfg.MinTimeDiff),
```
Comment (Dieterbe) [inline heap/heap.go]: why introduce a fake/incorrect value when we can simply rely on the zero value like we did before?

File: heap/heap.go [hunk 5]
```diff
@@ -14,43 +16,57 @@ import (
...
+		Errors:        errors,
+	}
+	proc, err := procfs.Self()
+	if err != nil {
+		heap.logError(err)
```
Comment (Dieterbe) [inline heap/heap.go]: the asynchronous feeding back of errors to the caller is for stuff that happens after construction time.
for errors known at construction time, we should just return them to the caller.
in particular, if we know at construction time that we won't be able to honor the requested threshold, the caller should know right then and there.

File: heap/heap.go [hunk 6]
```diff
@@ -14,43 +16,57 @@ import (
...
 	}
+
 	return &heap, nil
 }
 
 func (heap Heap) logError(err error) {
 	if heap.Errors != nil {
-		heap.Errors <- err
+		select {
+		case heap.Errors <- err:
```
Comment (Dieterbe) [inline heap/heap.go]: why silently drop errors? it's the caller responsibility to read the errors (if a non-nil chan was provided)

File: heap/heap.go [hunk 7]
```diff
@@ -14,43 +16,57 @@ import (
...
+	Errors        chan error
+
+	proc *procfs.Proc
+}
+
+// Config is the config for triggering profile
+type Config struct {
+	Path           string
+	AllocThreshold int
+	RSSThreshold   int
```
Comment (Dieterbe) [inline heap/heap.go]: perhaps this is personal preference, but when words are partially the same, i like to make that clearer and consistent by starting with the part that is identical.
e.g. i would name these
```
ThreshAlloc
ThreshRSS
```

File: heap/heap.go [hunk 8]
```diff
@@ -6,6 +6,8 @@ import (
 	"runtime"
 	"runtime/pprof"
 	"time"
+
+	"github.com/prometheus/procfs"
 )
 
 // Heap will check every checkEvery for memory obtained from the system, by the process
```
Comment (Dieterbe) [inline heap/heap.go]: the docs/comments need to be updated

File: heap/heap.go (hunk 7 above)
Comment (Dieterbe) [reply heap/heap.go]: also can you add a comment for both to describe whether they should be in bytes or in another unit.

**Review by Dieterbe — CHANGES_REQUESTED** [review]
needs changes

**Comment by Dieterbe** [comment]
@gouthamve any plans to finish this?

**Review by Dieterbe — APPROVED** [review]

**Pull request merged**